lib/web_ui/lib/src/engine/embedder.dart

```diff
@@ -180,102 +133,32 @@ class FlutterViewEmbedder {
-    final DomCSSStyleSheet sheet = _styleElement!.sheet! as DomCSSStyleSheet;
-    applyGlobalCssRulesToSheet(
-      sheet,
-      browserEngine: browserEngine,
-      hasAutofillOverlay: browserHasAutofillOverlay(),
-    );
-
-    final DomHTMLBodyElement bodyElement = domDocument.body!;
-
-    bodyElement.setAttribute(
+    _applicationDom.setHostAttribute(
       'flt-renderer',
       '${renderer.rendererTag} (${FlutterConfiguration.flutterWebAutoDetect ? 'auto-selected' : 'requested explicitly'})',
     );
-    bodyElement.setAttribute('flt-build-mode', buildMode);
-
-    setElementStyle(bodyElement, 'position', 'fixed');
-    setElementStyle(bodyElement, 'top', '0');
-    setElementStyle(bodyElement, 'right', '0');
-    setElementStyle(bodyElement, 'bottom', '0');
-    setElementStyle(bodyElement, 'left', '0');
-    setElementStyle(bodyElement, 'overflow', 'hidden');
-    setElementStyle(bodyElement, 'padding', '0');
-    setElementStyle(bodyElement, 'margin', '0');
-
-    // TODO(yjbanov): fix this when KVM I/O support is added. Currently scroll
-    //                using drag, and text selection interferes.
-    setElementStyle(bodyElement, 'user-select', 'none');
-    setElementStyle(bodyElement, '-webkit-user-select', 'none');
-    setElementStyle(bodyElement, '-ms-user-select', 'none');
-    setElementStyle(bodyElement, '-moz-user-select', 'none');
-
-    // This is required to prevent the browser from doing any native touch
-    // handling. If this is not done, the browser doesn't report 'pointermove'
-    // events properly.
-    setElementStyle(bodyElement, 'touch-action', 'none');
-
-    // These are intentionally outrageous font parameters to make sure that the
-    // apps fully specify their text styles.
-    setElementStyle(bodyElement, 'font', defaultCssFont);
-    setElementStyle(bodyElement, 'color', 'red');
-
+    _applicationDom.setHostAttribute('flt-build-mode', buildMode);
+    _applicationDom.setHostAttribute('flt-application-dom', _applicationDom.type);
     // TODO(mdebbar): Disable spellcheck until changes in the framework and
     // engine are complete.
-    bodyElement.spellcheck = false;
-
-    for (final DomElement viewportMeta
-        in domDocument.head!.querySelectorAll('meta[name="viewport"]')) {
-      if (assertionsEnabled) {
-        // Filter out the meta tag that the engine placed on the page. This is
-        // to avoid UI flicker during hot restart. Hot restart will clean up the
-        // old meta tag synchronously with the first post-restart frame.
-        if (!viewportMeta.hasAttribute('flt-viewport')) {
-          print(
-            'WARNING: found an existing <meta name="viewport"> tag. Flutter '
-            'Web uses its own viewport configuration for better compatibility '
-            'with Flutter. This tag will be replaced.',
-          );
-        }
-      }
-      viewportMeta.remove();
-    }
+    _applicationDom.setHostAttribute('spellcheck', 'false');
```

Comment:
This can also be set internally by the implementation. Together with the suggestions above, I think we can remove `setHostAttribute` too. I think it would be best for `ApplicationDom` to fully own its element, and not let random mutations of its style and other attributes from outside.

Reply:
I'll take a look at the API surface and simplify the public bits. We may only need a "prepareHostElement" method and some incoming configuration for it to work.